src/3d/qgsrubberband3d.cpp

```diff
@@ -94,11 +89,63 @@ QgsRubberBand3D::QgsRubberBand3D( Qgs3DMapSettings &map, QgsWindow3DEngine *engi
   mMarkerEntity->addComponent( mMarkerGeometryRenderer );
 }
 
+void QgsRubberBand3D::setupLine( Qt3DCore::QEntity *parentEntity, QgsWindow3DEngine *engine )
+{
+  mLineEntity = new Qt3DCore::QEntity( parentEntity );
+
+  QgsLineVertexData dummyLineData;
+  mLineGeometry = dummyLineData.createGeometry( mLineEntity );
+
+  Q_ASSERT( mLineGeometry->attributes().count() == 2 );
+  mPositionAttribute = mLineGeometry->attributes().at( 0 );
+  mIndexAttribute = mLineGeometry->attributes().at( 1 );
+
+  mLineGeometryRenderer = new Qt3DRender::QGeometryRenderer;
+  mLineGeometryRenderer->setPrimitiveType( Qt3DRender::QGeometryRenderer::LineStripAdjacency );
+  mLineGeometryRenderer->setGeometry( mLineGeometry );
+  mLineGeometryRenderer->setPrimitiveRestartEnabled( true );
+  mLineGeometryRenderer->setRestartIndexValue( 0 );
+
+  mLineEntity->addComponent( mLineGeometryRenderer );
+
+  mLineMaterial = new QgsLineMaterial;
+  mLineMaterial->setLineWidth( mWidth );
+  mLineMaterial->setLineColor( mColor );
+
+  QObject::connect( engine, &QgsAbstract3DEngine::sizeChanged, mLineMaterial, [this, engine] {
+    mLineMaterial->setViewportSize( engine->size() );
+  } );
+  mLineMaterial->setViewportSize( engine->size() );
+
+  mLineEntity->addComponent( mLineMaterial );
+}
+
+void QgsRubberBand3D::setupPolygon( Qt3DCore::QEntity *parentEntity )
+{
+  mPolygonEntity = new Qt3DCore::QEntity( parentEntity );
+
+  mPolygonGeometry = new QgsTessellatedPolygonGeometry();
+
+  mPolygonGeometryRenderer = new Qt3DRender::QGeometryRenderer;
+  mPolygonGeometryRenderer->setPrimitiveType( Qt3DRender::QGeometryRenderer::Triangles );
+  mPolygonGeometryRenderer->setGeometry( mPolygonGeometry );
+  mPolygonEntity->addComponent( mPolygonGeometryRenderer );
+
+  mPolygonMaterial = new QgsPhongMaterialSettings();
+  mPolygonMaterial->setAmbient( mColor );
+  mPolygonMaterial->setDiffuse( mColor );
+  mPolygonMaterial->setOpacity( 0.25 );
```

Comment:
this is a bit confusing, because `mPolygonMaterial` is not a _material_, it is just "material settings" object. I would suggest to do this instead:
```cpp
QgsPhongMaterialSettings materialSettings;
materialSettings.setAmbient( mColor );
// ...
mPolygonMaterial = materialSettings.toMaterial( ... );   // mPolygonMaterial is QgsMaterial *
mPolygonEntity->addComponent( mPolygonMaterial );
```

This should then also make things work correctly when you want to remove the material component later